pam/go-loader/module.c

```diff
@@ -0,0 +1,142 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General
+ * Public License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Marco Trevisan <marco.trevisan@canonical.com>
+ */
+
+#include <dlfcn.h>
+#include <limits.h>
+#include <security/pam_modules.h>
+#include <security/pam_ext.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+
+/* When a Go shared library is loaded from C, go starts various goroutine
+ * (as init() at first) and if the loading code is then performing a fork
+ * we end up having an undefined behavior and very likely, deadlocks.
+ *
+ * To avoid this we need to ensure that the module is called after the loading
+ * application has forked, and this for sure has happened when the application
+ * calls the PAM functions that a module exposes.
+ *
+ * As per this, here we re-implement the PAM modules functions and for each
+ * action we load the module if it has not been loaded already, otherwise we
+ * dload it and we redirect each call to the loaded library.
+ */
+
+typedef int (*PamHandler)(pam_handle_t *,
+                          int          flags,
+                          int          argc,
+                          const char **argv);
+
+static void
+on_go_module_removed (pam_handle_t *pamh,
+                      void         *go_module,
+                      int           error_status)
+{
+  void (*go_pam_cleanup) (void);
+  *(void **) (&go_pam_cleanup) = dlsym (go_module, "go_pam_cleanup_module");
+  if (go_pam_cleanup)
+    go_pam_cleanup ();
+
+  dlclose (go_module);
+}
+
+static void *
+load_module (pam_handle_t *pamh,
+             const char   *module_path)
+{
+  void *go_module;
+
+  if (pam_get_data (pamh, "go-module", (const void **) &go_module) == PAM_SUCCESS)
+    return go_module;
+
+  go_module = dlopen (module_path, RTLD_LAZY);
+  if (!go_module)
+    return NULL;
+
+  pam_set_data (pamh, "go-module", go_module, on_go_module_removed);
+
+  void (*init_module) (void);
+  *(void **) (&init_module) = dlsym (go_module, "go_pam_init_module");
+  if (init_module)
+    init_module ();
+
+  return go_module;
+}
+
+static inline int
+call_pam_function (pam_handle_t *pamh,
+                   const char   *function,
+                   int           flags,
+                   int           argc,
+                   const char  **argv)
+{
+  char module_path[PATH_MAX] = {0};
+  const char *sub_module;
+  PamHandler func;
+  void *go_module;
+
+  if (argc < 1)
+    {
+      pam_error (pamh, "%s: no module provided", function);
+      return PAM_MODULE_UNKNOWN;
+    }
+
+  sub_module = argv[0];
+  argc -= 1;
+  argv = (argc == 0) ? NULL : &argv[1];
+
+  snprintf (module_path, PATH_MAX - 1, PAM_MODULES_DIR "/%s", sub_module);
+
+#if DEBUG_MODE
```

Comment:
Not a big fan of this precompilation, especially as you are taking the module (and so its path) already as an argument. We can just change the PAM stack to reflect this, no?

Reply:
Here the point is:
 - If in `DEBUG_MODE`, you can set a library in *any* path to be loaded (e.g. currently my pam service file is `auth	required        /home/marco/Dev/authd/_build/pam_go_loader.so /home/marco/Dev/authd/pam/pam_authd.so socket=/tmp/authd.sock`)
 - When in release mode, instead the library can be loaded from a trusted path i.e. `/lib/x86_64-linux-gnu/security/` (indeed the check has to be improved to not allow `..` or `/` in the loaded path